[CHAPTER] Add chapter 12 on fine-tuning reasoning models (#799)
huggingface/course: chapters/en/chapter13/2.mdx

Reinforcement Learning from Human Feedback has been used to train many of the most popular LLMs today, such as OpenAI's GPT-4, Google's Gemini, and DeepSeek's R1. There are a wide range of techniques for RLHF, with varying degrees of complexity and sophistication. In this chapter, we will focus on Generalized Policy Rectification Optimization (GRPO), which is a technique for RLHF that has been shown to be effective at training LLMs that are helpful, harmless, and aligned with human preferences.

## Why should we care about GRPO (Generalized Policy Rectification Optimization)?

GRPO (Generalized Policy Rectification Optimization) represents a significant advancement in reinforcement learning for language models. It was developed by DeepSeek and played a crucial role in training their DeepSeek-R1 model, which has achieved state-of-the-art performance on various reasoning benchmarks.

This course is focused on GRPO and its implementation in the DeepSeek-R1 model. But before we dive into the technical details of GRPO, let's look at the benefits of using GRPO over other RLHF techniques.

> | Aspect | Description | Technical Significance |

Author: Sorry to waste your time Ed. I'll redo this section because It should basically just explain why the course focuses on GRPO opposed to other algorithms.

|--------|-------------|------------------------|

> | Direct Preference Learning | GRPO directly learns from preference data without requiring explicit reward modeling | Eliminates the need for a separate reward model, reducing complexity and potential error sources in the training pipeline |

Reviewer: I think you are thinking of DPO here, GRPO requires an external verifier for Math / code and an RM for more ambiguous tasks such as creative writing.

Author: Got it. Thanks. 

I was thinking that in its minimal format (for example [length based function](https://huggingface.co/docs/trl/main/en/grpo_trainer)) it doesn't require an extra model. 

What do you think about modifying this row to make it clear I'm talking about the minimum, and adding a new row to mention tasks like creative writing.

> | Enhanced Stability | Uses a novel group-based optimization approach that considers multiple responses simultaneously | The group-based approach provides more stable gradients and better convergence properties compared to traditional pairwise comparisons |

Reviewer: I am not sure if makes sense to mention pairwise comparisons here. Maybe rephrase as "The group-based approach provides more stable gradients and better convergence properties compared to the typical pointwise RL methods"

| Computational Efficiency | Achieves better sample efficiency through its group-based training mechanism | Reduces the computational resources needed for training, making it more practical for large-scale language models |
| Scalable Architecture | Designed specifically for large language models with distributed training in mind | Enables effective training of models with billions of parameters while maintaining stability |
| Improved Reasoning | Particularly effective at enhancing multi-step reasoning capabilities | The group optimization approach helps the model learn better reasoning patterns by considering multiple solution paths simultaneously |
| Training Pipeline Integration | Seamlessly integrates with existing pre-training and fine-tuning workflows | Can be applied as a post-training step without requiring changes to the base model architecture |

GRPO addresses several key limitations of previous approaches like PPO (Proximal Policy Optimization):

> 1. **Group vs Pairwise**: Instead of comparing just two responses, GRPO can optimize across groups of responses simultaneously, leading to more robust learning.

Reviewer: I am not sure why this pairwise stuff keeps coming back in. DPO works with pairwise data. PPO you generate one sample per prompt, score with an RM and perform optimization (over batches of different prompts). For GRPO n generations are created per prompt, scored with an RM / verifier and the scores as used as a baseline estimate. One advantage of GRPO over PPO is that a critic model is not required, as this is replace by the baseline estimate.

Author: Here, I think I mean 'sample/point wise' rather than pair. i.e. GRPO compares accross a group of samples rather than a single one, like other algorithms.

However, just being fluent isn't enough. We want our LLMs to be more than just good at stringing words together. We want them to be:

* **Helpful:** Provide useful and relevant information.  
* **Harmless:** Avoid generating toxic, biased, or harmful content.  
* **Aligned with Human Preferences:** Respond in ways that humans find natural, helpful, and engaging.

Pre-training LLM methods, which mostly rely on predicting the next word from text data, sometimes fall short on these aspects.

Whilst supervised training is excellent at producing structured outputs, it can be less effective at producing helpful, harmless, and aligned responses. We explore supervised training in [chapter 11](/chapters/en/chapter11/1).

Fine-tuned models might generate fluent and structured text that is still factually incorrect, biased, or doesn't really answer the user's question in a helpful way.

**Enter Reinforcement Learning\!** RL gives us a way to fine-tune these pre-trained LLMs to better achieve these desired qualities. It's like giving our LLM dog extra training to become a well-behaved and helpful companion, not just a dog that knows how to bark fluently\!

> ## Reinforcement Learning from Human Feedback (RLHF)

Reviewer 2: do you think we can make a diagram here to showcase these too - RLHF, DPO, PPO?

# Introduction to Reinforcement Learning and its Role in LLMs

Welcome to the first page! 

We're going to start our journey into the exciting world of Reinforcement Learning (RL) and discover how it's revolutionizing the way we train Language Models like the ones you might use every day.

This page will give you a friendly and clear introduction to RL, even if you've never encountered it before. We'll break down the core ideas and see why RL is becoming so important in the field of Large Language Models (LLMs).

## What is Reinforcement Learning (RL)?

Imagine you're training a dog. You want to teach it to sit. You might say "Sit!" and then, if the dog sits, you give it a treat and praise. If it doesn't sit, you might gently guide it or just try again. Over time, the dog learns to associate sitting with the positive reward (treat and praise) and is more likely to sit when you say "Sit!" again.

> That, in a nutshell, is the basic idea behind Reinforcement Learning! Instead of a dog, we have a **language model** (in reinforcement learning, we call it an **agent**), and instead of you, we have the **environment** that gives feedback.

Reviewer 3: you can add that we call the feedback the _reward_

> ![RL Basic Concept](https://huggingface.co/nlp-course/images/resolve/main/grpo/2.jpg)  

Reviewer 3: IMO, this image is misleading. GRPO, while compatible with reward model is in the vast majority of cases used with a static reward function. And there is no such thing as preference here

Author: I changes the image to reflect that rewards can come from different sources but represent human preferences.

![updates](https://huggingface.co/nlp-course/images/resolve/main/grpo/2.jpg)

I don't want to be GRPO specific at this point in the material, because we're introducing RL generally.

### Reward

This is the feedback the environment gives to the agent after it takes an action. Rewards are usually numbers. 

**Positive rewards** are like treats and praise – they tell the agent "good job, you did something right!". 

**Negative rewards** (or penalties) are like a gentle "no" – they tell the agent "that wasn't quite right, try something else". For the dog, the treat is the reward. 

For an LLM, rewards are designed to reflect how well the LLM is doing at a specific task – maybe it's how helpful, truthful, or harmless its response is.

### Policy

This is the agent's strategy for choosing actions. It's like the dog's understanding of what it should do when you say "Sit!". In RL, the policy is what we're really trying to learn and improve. It's a set of rules or a function that tells the agent what action to take in different situations. Initially, the policy might be random, but as the agent learns, the policy becomes better at choosing actions that lead to higher rewards.

> ## The RL Learning Process: Trial and Error

Reviewer 3: ```suggestion
## The RL Process: Trial and Error
```

![RL Process](https://huggingface.co/nlp-course/images/resolve/main/grpo/1.jpg)

Reinforcement Learning happens through a process of trial and error:

| Step | Process | Description |
|------|---------|-------------|
| 1. Observation | The agent observes the environment | The agent takes in information about its current state and surroundings |
| 2. Action | The agent takes an action based on its current policy | Using its learned strategy (policy), the agent decides what to do next |
| 3. Feedback | The environment gives the agent a reward (or penalty) | The agent receives feedback on how good or bad its action was |

> | 4. Learning | The agent updates its policy based on the reward | The agent adjusts its strategy - reinforcing actions that led to good rewards and avoiding those that led to penalties |

Reviewer 3: ```suggestion
| 4. Learning | The agent updates its policy based on the reward | The agent adjusts its strategy - reinforcing actions that led to high rewards and avoiding those that led low rewards |
```

* **Aligned with Human Preferences:** Respond in ways that humans find natural, helpful, and engaging.

Pre-training LLM methods, which mostly rely on predicting the next word from text data, sometimes fall short on these aspects.

Whilst supervised training is excellent at producing structured outputs, it can be less effective at producing helpful, harmless, and aligned responses. We explore supervised training in [chapter 11](/chapters/en/chapter11/1).

Fine-tuned models might generate fluent and structured text that is still factually incorrect, biased, or doesn't really answer the user's question in a helpful way.

**Enter Reinforcement Learning\!** RL gives us a way to fine-tune these pre-trained LLMs to better achieve these desired qualities. It's like giving our LLM dog extra training to become a well-behaved and helpful companion, not just a dog that knows how to bark fluently\!

## Reinforcement Learning from Human Feedback (RLHF)

A very popular technique for aligning language models is **Reinforcement Learning from Human Feedback (RLHF)**. In RLHF, we use human feedback as a proxy for the "reward" signal in RL. Here's how it works:

> ![RLHF Process](https://huggingface.co/nlp-course/images/resolve/main/grpo/3.jpg)

Reviewer 3: where's human in the image?

1. **Get Human Preferences:** We might ask humans to compare different responses generated by the LLM for the same input prompt and tell us which response they prefer. For example, we might show a human two different answers to the question "What is the capital of France?" and ask them "Which answer is better?".

2. **Train a Reward Model:** We use this human preference data to train a separate model called a **reward model**. This reward model learns to predict what kind of responses humans will prefer. It learns to score responses based on helpfulness, harmlessness, and alignment with human preferences.

> 3. **Fine-tune the LLM with RL:** Now we use the reward model as the environment for our LLM agent. The LLM generates responses (actions), and the reward model scores these responses (provides rewards). In essence, we're training the LLM to produce text that our reward model (which learned from human preferences) thinks is good.

Reviewer 3: ```suggestion
3. **Fine-tune the LLM with RL:** Now we use the reward model as the environment for our LLM agent. The LLM generates responses (actions), and the reward model scores these responses. In essence, we're training the LLM to produce text that our reward model (which learned from human preferences) thinks is good.
```

not necessarily positive

Reviewer 3: Sorry I read "positive" forget about it

From a general perspective, let's look at the benefits of using RL in LLMs:

| Benefit | Description |
|---------|-------------|
| Improved Control | RL allows us to have more control over the kind of text LLMs generate. We can guide them to produce text that is more aligned with specific goals, like being helpful, creative, or concise. |
| Enhanced Alignment with Human Values | RLHF, in particular, helps us align LLMs with complex and often subjective human preferences. It's hard to write down rules for "what makes a good answer," but humans can easily judge and compare responses. RLHF lets the model learn from these human judgments. |
| Mitigating Undesirable Behaviors | RL can be used to reduce negative behaviors in LLMs, such as generating toxic language, spreading misinformation, or exhibiting biases. By designing rewards that penalize these behaviors, we can nudge the model to avoid them. |

> Reinforcement Learning from Human Feedback has been used to train many of the most popular LLMs today, such as OpenAI's GPT-4, Google's Gemini, and DeepSeek's R1. There are a wide range of techniques for RLHF, with varying degrees of complexity and sophistication. In this chapter, we will focus on Generalized Policy Rectification Optimization (GRPO), which is a technique for RLHF that has been shown to be effective at training LLMs that are helpful, harmless, and aligned with human preferences.

Reviewer 3: > Generalized Policy Rectification Optimization (GRPO)

your LLM doesn't know about GRPO 😂

## Why should we care about GRPO (Generalized Policy Rectification Optimization)?

There are many techniques for RLHF but this course is focused on GRPO because it represents a significant advancement in reinforcement learning for language models.

Let's briefly consider two of other popular techniques for RLHF:

- Proximal Policy Optimization (PPO)
- Direct Preference Optimization (DPO)

> Proximal Policy Optimization (PPO) was one of the first highly effective techniques for RLHF. It uses a policy gradient method to update the policy based on the reward from a separate reward model.

Reviewer 3: another link to RL course can be relevant here

Direct Preference Optimization (DPO) was later developed as a simpler technique that eliminates the need for a separate reward model using preference data directly. Essentially, framing the problem as a classification task between the chosen and rejected responses.

<Tip>

DPO and PPO are complex reinforcement learning algorithms in their own right, which we will not cover in this course. If you're interested in learning more about them, you can check out the following resources:

- [Proximal Policy Optimization](https://huggingface.co/docs/trl/main/en/ppo_trainer)
- [Direct Preference Optimization](https://huggingface.co/docs/trl/main/en/dpo_trainer)

</Tip>

> Unlike DPO and PPO, GRPO groups similar samples together and compares them as a group. The group-based approach provides more stable gradients and better convergence properties compared to other methods.

Reviewer 3: I don't know if it's said later, but here I think it's important to mention that in GRPO, we don't have the stages that were introduces earlier "collect human preferences" and "train a reward model"

Reviewer 3: ok it's literally this next line 😂

For example, we could use a length function to reward shorter responses, a mathematical solver to verify solution correctness, or a factual correctness function to reward responses that are more factually accurate. This flexibility makes GRPO particularly versatile for different types of alignment tasks.

### Comparing RLHF Methods: GRPO vs PPO vs DPO

To better understand GRPO's advantages, here's a comparison with other popular RLHF methods:

| Feature | GRPO | PPO | DPO |
|---------|------|-----|-----|
| **Core Approach** | Group-based optimization with normalized rewards | Policy gradient with clipped objective | Direct preference optimization |
| **Reward Source** | Any function or model | Separate reward model | Derived from preference data |
| **Implementation Complexity** | Medium | High | Low |

This comparison highlights why GRPO represents an important advancement in RLHF techniques, offering a balance of efficiency, flexibility, and stability.

> This is a very minimal overview of GRPO. In the next section, we'll dive into the details of GRPO and see how it works in practice.

Reviewer 3: Since PPO and DPO aren't covered, I'm not convinced that this is relevant
